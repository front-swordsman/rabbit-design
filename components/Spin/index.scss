@import '../styles/themes/default';

$spin: 'snake-spin';

.#{$spin} {
  font-size: $body;
  display: inline-block;
  position: relative;

  &-wrapper {
    .#{$spin}-loading {
      position: absolute;
      height: 100%;
      width: 100%;
      z-index: 2;
      background: rgba(#fff, 0.7);

      &-container {
        position: absolute;
        left: 50%;
        top: 50%;
        transform: translate(-50%, -50%);
      }
    }
  }

  &-loading {
    &-icon {
      display: block;
      text-align: center;
    }

    &-text {
      color: $brand;
      text-align: center;
    }
  }

  &-sk-child {
    width: 100%;
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;

    &::before {
      content: '';
      display: block;
      margin: 0 auto;
      width: 15%;
      height: 15%;
      background-color: $brand;
      border-radius: 100%;
      animation: sk-circleBounceDelay 1.2s infinite ease-in-out both;
    }
  }

  &-sk-circle2 {
    transform: rotate(30deg);

    &::before {
      animation-delay: -1.1s;
    }
  }

  &-sk-circle3 {
    transform: rotate(60deg);

    &::before {
      animation-delay: -1s;
    }
  }

  &-sk-circle4 {
    transform: rotate(90deg);

    &::before {
      animation-delay: -0.9s;
    }
  }

  &-sk-circle5 {
    transform: rotate(120deg);

    &::before {
      animation-delay: -0.8s;
    }
  }

  &-sk-circle6 {
    transform: rotate(150deg);

    &::before {
      animation-delay: -0.7s;
    }
  }

  &-sk-circle7 {
    transform: rotate(180deg);

    &::before {
      animation-delay: -0.6s;
    }
  }

  &-sk-circle8 {
    transform: rotate(210deg);

    &::before {
      animation-delay: -0.5s;
    }
  }

  &-sk-circle9 {
    transform: rotate(240deg);

    &::before {
      animation-delay: -0.4s;
    }
  }

  &-sk-circle10 {
    transform: rotate(270deg);

    &::before {
      animation-delay: -0.3s;
    }
  }

  &-sk-circle11 {
    transform: rotate(300deg);

    &::before {
      animation-delay: -0.2s;
    }
  }

  &-sk-circle12 {
    transform: rotate(330deg);

    &::before {
      animation-delay: -0.1s;
    }
  }

  &-small {
    .#{$spin}-default-spin {
      width: 24px;
      height: 24px;
    }
  }

  &-default-spin {
    display: inline-block;
    width: 32px;
    height: 32px;
    position: relative;
    margin: 0;
  }
}

@keyframes sk-circleBounceDelay {
  0%,
  80%,
  100% {
    transform: scale(0);
  }

  40% {
    transform: scale(1);
  }
}
