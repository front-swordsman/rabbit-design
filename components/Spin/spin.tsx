import * as React from 'react'
import cx from 'classnames'
import { SpinProps } from 'types/spin'

import './index.scss'

const defaultProps = {
  size: 'normal'
}

const prefixCls = 'snake-spin'

const Spin: React.FC<SpinProps> = (spin: SpinProps, ref: any) => {
  const props = { ...defaultProps, ...spin }
  const { size, children, indicator, tip } = props

  const getClassNames = () => {
    return cx(prefixCls, {
      [`${prefixCls}-${size}`]: size,
      [`${prefixCls}-wrapper`]: children
    })
  }

  const renderIndicator = () => {
    return (
      <div className={`${prefixCls}-loading-icon`}>
        {indicator || (
          <div className={`${prefixCls}-default-spin`}>
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle1`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle2`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle3`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle4`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle5`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle6`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle7`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle8`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle9`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle10`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle11`} />
            <div className={`${prefixCls}-sk-child ${prefixCls}-sk-circle12`} />
          </div>
        )}
      </div>
    )
  }

  const renderChildren = () => {
    if (children) {
      return <div className={`${prefixCls}-container`}>{children}</div>
    }
    return null
  }

  const renderComponent = () => {
    const classStr = getClassNames()
    return (
      <div className={classStr} ref={ref}>
        <div className={`${prefixCls}-loading`}>
          <div className={`${prefixCls}-loading-container`}>
            {renderIndicator()}
            {tip ? <div className={`${prefixCls}-loading-text`}>{tip}</div> : null}
          </div>
        </div>
        {renderChildren()}
      </div>
    )
  }

  return renderComponent()
}

export default React.forwardRef(Spin)
