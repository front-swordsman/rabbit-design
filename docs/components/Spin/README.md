---
imports:
  import SimpleDemoCode from '!raw-loader!./simpleDemo.tsx';
  import SimpleDemo from './simpleDemo.tsx';
  import SizeDemoCode from '!raw-loader!./sizeDemo.tsx';
  import SizeDemo from './sizeDemo.tsx';
  import TipDemoCode from '!raw-loader!./tipDemo.tsx';
  import TipDemo from './tipDemo.tsx';
  import ContainerDemoCode from '!raw-loader!./container.tsx';
  import ContainerDemo from './container.tsx';
---

# Spin 加载中

用于页面和区块的加载中状态。

## 简单示例

:::demo
<Block des="简单的 spin" code={SimpleDemoCode}>
  <SimpleDemo />
</Block>
:::

## 多尺寸示例

:::demo
<Block des="多尺寸 spin" code={SizeDemoCode}>
  <SizeDemo />
</Block>
:::

## 文字提示

:::demo
<Block des="文字提示 spin" code={TipDemoCode}>
  <TipDemo />
</Block>
:::

## 区块加载

:::demo
<Block des="区块加载 spin" code={ContainerDemoCode}>
  <ContainerDemo />
</Block>
:::

## Api

| 参数      | 说明             | 类型              | 可选值 | 默认值   |
| --------- | ---------------- | ----------------- | ------ | -------- |
| indicator | 自定义的指示器   | ReactNode         | --     | --       |
| size      | 尺寸             | `small`, `normal` | --     | `normal` |
| tip       | 自定义的文字提示 | ReactNode         | --     | --       |