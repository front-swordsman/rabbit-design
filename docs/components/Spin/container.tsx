import * as React from 'react'
import { Spin } from 'components'

export default function Container() {
  return (
    <div>
      <Spin tip="loading">
        <div style={{ width: 300 }}>
          内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区
          内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区
          内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区
          内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区
          内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区
        </div>
      </Spin>
    </div>
  )
}
