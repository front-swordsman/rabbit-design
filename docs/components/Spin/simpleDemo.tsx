import * as React from 'react'
import { Spin } from 'components'

export default function SimpleDemo() {
  return <Spin />
}
