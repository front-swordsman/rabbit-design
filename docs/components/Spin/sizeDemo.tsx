import * as React from 'react'
import { Spin } from 'components'

export default function SizeDemo() {
  return (
    <div>
      <div>
        <Spin size="normal" />
        <div>正常尺寸</div>
      </div>
      <div>
        <Spin size="small" />
        <div>小尺寸</div>
      </div>
    </div>
  )
}
