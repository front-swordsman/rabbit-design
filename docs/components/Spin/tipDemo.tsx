import * as React from 'react'
import { Spin } from 'components'

export default function TipDemo() {
  return (
    <div>
      <Spin tip="loading" />
    </div>
  )
}
