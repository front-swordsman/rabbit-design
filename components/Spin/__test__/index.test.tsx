import * as React from 'react'
import { render } from 'react-testing-library'
import Spin from '../index'

describe('spin test', () => {
  it('snapshot with size equal normal', () => {
    const { container } = render(<Spin />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with size equal small', () => {
    const { container } = render(<Spin size="small" />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('snapshot with children', () => {
    const { container } = render(
      <Spin>内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区</Spin>
    )

    expect(container.innerHTML).toMatchSnapshot()
  })

  it('spin with tip', () => {
    const { getByText } = render(<Spin tip="loading" />)
    expect(getByText(/loading/i)).not.toBeNull()
  })
})
