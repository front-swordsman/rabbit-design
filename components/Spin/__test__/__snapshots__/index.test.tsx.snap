// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spin test snapshot with children 1`] = `"<div class=\\"snake-spin snake-spin-normal snake-spin-wrapper\\"><div class=\\"snake-spin-loading\\"><div class=\\"snake-spin-loading-container\\"><div class=\\"snake-spin-loading-icon\\"><div class=\\"snake-spin-default-spin\\"><div class=\\"snake-spin-sk-child snake-spin-sk-circle1\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle2\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle3\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle4\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle5\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle6\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle7\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle8\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle9\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle10\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle11\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle12\\"></div></div></div></div></div><div class=\\"snake-spin-container\\">内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区内容区</div></div>"`;

exports[`spin test snapshot with size equal normal 1`] = `"<div class=\\"snake-spin snake-spin-normal\\"><div class=\\"snake-spin-loading\\"><div class=\\"snake-spin-loading-container\\"><div class=\\"snake-spin-loading-icon\\"><div class=\\"snake-spin-default-spin\\"><div class=\\"snake-spin-sk-child snake-spin-sk-circle1\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle2\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle3\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle4\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle5\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle6\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle7\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle8\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle9\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle10\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle11\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle12\\"></div></div></div></div></div></div>"`;

exports[`spin test snapshot with size equal small 1`] = `"<div class=\\"snake-spin snake-spin-small\\"><div class=\\"snake-spin-loading\\"><div class=\\"snake-spin-loading-container\\"><div class=\\"snake-spin-loading-icon\\"><div class=\\"snake-spin-default-spin\\"><div class=\\"snake-spin-sk-child snake-spin-sk-circle1\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle2\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle3\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle4\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle5\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle6\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle7\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle8\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle9\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle10\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle11\\"></div><div class=\\"snake-spin-sk-child snake-spin-sk-circle12\\"></div></div></div></div></div></div>"`;
